File: bestdog/scratch/sort_changelog.py
import re

def sort_changelog(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    # Split by version headers
    # Regex to match "## [version] - date" and capture the version
    blocks = re.split(r'(?m)^## \[(.*?)\]', content)
    
    header = blocks[0]
    versions_and_content = []
    
    for i in range(1, len(blocks), 2):
        version = blocks[i]
        # Content is the rest until the next version
        raw_content = blocks[i+1]
        versions_and_content.append((version, raw_content))

    # Sort versions in descending order
    def version_key(v):
        # Convert "4.2.5" to [4, 2, 5]
        return [int(x) for x in v.split('.')]

    sorted_entries = sorted(versions_and_content, key=lambda x: version_key(x[0]), reverse=True)

    # Reconstruct content
    new_content = header
    for version, content in sorted_entries:
        new_content += f"## [{version}]{content}"

    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(new_content)

if __name__ == "__main__":
    sort_changelog('CHANGELOG.md')

File: bestdog/scratch/sort_changelog.js
const fs = require('fs');

function sortChangelog(filepath) {
    const content = fs.readFileSync(filepath, 'utf8');
    
    // Split by version headers using regex
    // We use a regex that matches the header and keeps the header in the results
    const regex = /^## \[(.*?)\]/gm;
    let match;
    const entries = [];
    let lastIndex = 0;
    let lastVersion = null;

    while ((match = regex.exec(content)) !== null) {
        if (lastVersion !== null) {
            entries.push({
                version: lastVersion,
                content: content.slice(lastIndex, match.index)
            });
        }
        lastVersion = match[1];
        lastIndex = match.index + match[0].length + match[1].length + 4; // approximate offset for the version and brackets
        // Wait, the regex.exec match.index points to the start of the match "## [version]"
        // Let's do it simpler.
    }
    
    // Manual split
    const parts = content.split(/^## \[/gm);
    const header = parts[0];
    const versionsAndContent = [];

    for (let i = 1; i < parts.length; i++) {
        const part = parts[i];
        const versionMatch = part.match(/^(.*?)\]/);
        if (versionMatch) {
            const version = versionMatch[1];
            const entryContent = part.slice(version.length + 1);
            versionsAndContent.push({ version, entryContent });
        }
    }

    function versionToKey(v) {
        return v.split('.').map(x => parseInt(x, 10));
    }

    function compareVersions(a, b) {
        const keyA = versionToKey(a);
        const keyB = versionToKey(b);
        for (let i = 0; i < Math.max(keyA.length, keyB.length); i++) {
            const valA = keyA[i] || 0;
            const valB = keyB[i] || 0;
            if (valA !== valB) return valB - valA; // Descending
        }
        return 0;
    }

    versionsAndContent.sort((a, b) => compareVersions(a.version, b.version));

    let newContent = header;
    for (const entry of versionsAndContent) {
        newContent += `## [${entry.version}]${entry.entryContent}`;
    }

    fs.writeFileSync(filepath, newContent, 'utf8');
}

sortChangelog('CHANGELOG.md');
